File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] max = new int[1];

    public int diameterOfBinaryTree(TreeNode root) {
        int i = recursiveDiameter(root,0);
        return max[0];
    }

    public int recursiveDiameter(TreeNode node, int level) {
        if(node.left == null && node.right == null) {
            return level;
        }
        int ml = level, mr = level;
        if(node.left != null) {
            ml = recursiveDiameter(node.left, level + 1);
        }
        if(node.right != null) {
            mr = recursiveDiameter(node.right, level + 1);
        }
        max[0] = Math.max(max[0],(ml - level) + (mr - level));

        return Math.max(ml,mr);
    }
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution(object):
    max_diameter = [0]

    def diameterOfBinaryTree(self, root):
        self.max_diameter[0] = 0  # Reset max_diameter for each call
        i = self.recursive_diameter(root, 0)
        return self.max_diameter[0]

    # Recursive function to calculate diameter
    def recursive_diameter(self, node, level):
        if node is None:
            return level
        ml = level
        mr = level
        if node.left is not None:
            ml = self.recursive_diameter(node.left, level + 1)
        if node.right is not None:
            mr = self.recursive_diameter(node.right, level + 1)
        self.max_diameter[0] = max(self.max_diameter[0], (ml - level) + (mr - level))
        return max(ml, mr)
    